File: Programs/Core/ModuleKind.go
// Copyright mzoesch. All rights reserved.

package Core

import "fmt"

type ModuleKind int

const (
    MODULE_KIND_SHARED ModuleKind = iota
    MODULE_KIND_STATIC ModuleKind = iota
    MODULE_KIND_LAUNCH ModuleKind = iota
)

func (kind ModuleKind) IsShared() bool {
    return kind == MODULE_KIND_SHARED
}

func (kind ModuleKind) IsStatic() bool {
    return kind == MODULE_KIND_STATIC
}

func (kind ModuleKind) IsLaunch() bool {
    return kind == MODULE_KIND_LAUNCH
}

func (kind ModuleKind) ToString() string {
    switch kind {
    case MODULE_KIND_SHARED:
        return "Shared"
    case MODULE_KIND_STATIC:
        return "Static"
    case MODULE_KIND_LAUNCH:
        return "Launch"
    default:
        panic("Unknown ModuleKind value.")
    }
}

func (kind ModuleKind) ToLuaString() string {
    switch kind {
    case MODULE_KIND_SHARED:
        return "SharedLib"
    case MODULE_KIND_STATIC:
        return "StaticLib"
    case MODULE_KIND_LAUNCH:
        return "ConsoleApp"
    default:
        panic("Unknown ModuleKind value.")
    }
}

func (kind ModuleKind) ToCMakeString() string {
    switch kind {
    case MODULE_KIND_SHARED:
        return "SHARED"
    case MODULE_KIND_STATIC:
        return "STATIC"
    default:
        panic("Unknown ModuleKind value.")
    }
}

func ModuleKindFromString(str string) ModuleKind {
    switch str {
    case "Shared", "SHARED", "SharedLib", "SHARED_LIBRARY":
        return MODULE_KIND_SHARED
    case "Static", "STATIC", "StaticLib", "STATIC_LIBRARY":
        return MODULE_KIND_STATIC
    case "Launch", "LAUNCH", "ConsoleApp", "EXECUTABLE":
        return MODULE_KIND_LAUNCH
    default:
        panic(fmt.Sprintf("Unknown ModuleKind value: [%s].", str))
    }
}

File: Motor/module_kind.rs
// Copyright mzoesch. All rights reserved.

#[derive(Debug, Clone, clap::ValueEnum)]
#[clap(rename_all = "lowercase")]
pub enum ModuleKind
{
    Unknown,
    Shared,
    Static,
    Launch,
    Plugin,
}

impl ModuleKind
{
    #[allow(dead_code)]
    pub fn is_shared_weak(&self) -> bool
    {
        match self
        {
            ModuleKind::Shared => true,
            ModuleKind::Plugin => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn is_unknown(&self) -> bool
    {
        match self
        {
            ModuleKind::Unknown => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn is_shared(&self) -> bool
    {
        match self
        {
            ModuleKind::Shared => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn is_static(&self) -> bool
    {
        match self
        {
            ModuleKind::Static => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn is_launch(&self) -> bool
    {
        match self
        {
            ModuleKind::Launch => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn is_plugin(&self) -> bool
    {
        match self
        {
            ModuleKind::Plugin => true,
            _ => false,
        }
    }

    #[allow(dead_code)]
    pub fn from_str(kind: &str) -> Option<ModuleKind>
    {
        return match kind
        {
            "Unknown" => Some(ModuleKind::Unknown),
            "Shared" => Some(ModuleKind::Shared),
            "SharedLib" => Some(ModuleKind::Shared),
            "SHARED_LIBRARY" => Some(ModuleKind::Shared),
            "Static" => Some(ModuleKind::Static),
            "StaticLib" => Some(ModuleKind::Static),
            "STATIC_LIBRARY" => Some(ModuleKind::Static),
            "Launch" => Some(ModuleKind::Launch),
            "ConsoleApp" => Some(ModuleKind::Launch),
            "EXECUTABLE" => Some(ModuleKind::Launch),
            "Plugin" => Some(ModuleKind::Plugin),
            _ => None,
        }
    }
}
